rust/lance-index/src/scalar/ngram.rs

```diff
@@ -0,0 +1,456 @@
+
+const TOKENS_COL: &str = "tokens";
+const POSTING_LIST_COL: &str = "posting_list";
+const POSTINGS_FILENAME: &str = "ngram_postings.lance";
+
+lazy_static::lazy_static! {
+    pub static ref TOKENS_FIELD: Field = Field::new(TOKENS_COL, DataType::Utf8, false);
+    pub static ref POSTINGS_FIELD: Field = Field::new(POSTING_LIST_COL, DataType::Binary, false);
+    pub static ref POSTINGS_SCHEMA: SchemaRef = Arc::new(Schema::new(vec![TOKENS_FIELD.clone(), POSTINGS_FIELD.clone()]));
+    /// Currently we ALWAYS use trigrams with ascii folding and lower casing.  We may want to make this configurable in the future.
+    pub static ref NGRAM_TOKENIZER: TextAnalyzer = TextAnalyzer::builder(tantivy::tokenizer::NgramTokenizer::all_ngrams(1, 3).unwrap())
+    .filter(tantivy::tokenizer::LowerCaser)
+    .filter(tantivy::tokenizer::AlphaNumOnlyFilter)
+    .filter(tantivy::tokenizer::AsciiFoldingFilter)
+    .build();
+}
+trait TextAnalyzerExt {
+    fn tokenize(self, text: &str) -> Vec<String>;
+}
+
+impl TextAnalyzerExt for TextAnalyzer {
+    fn tokenize(mut self, text: &str) -> Vec<String> {
+        let mut stream = self.token_stream(text);
+        let mut tokens = Vec::new();
+        while stream.advance() {
+            let token = stream.token();
+            tokens.push(token.text.clone());
+        }
+        tokens
+    }
+}
+
+/// Basic stats about an ngram index
+#[derive(Serialize)]
+struct NGramStatistics {
+    num_ngrams: usize,
+}
+
+/// The row ids that contain a given ngram
+#[derive(Debug)]
+struct NGramPostingList {
+    bitmap: RoaringTreemap,
+}
+
+impl DeepSizeOf for NGramPostingList {
+    fn deep_size_of_children(&self, _: &mut deepsize::Context) -> usize {
+        self.bitmap.serialized_size() as usize
+    }
+}
+
+impl NGramPostingList {
+    fn try_from_batch(batch: RecordBatch) -> Result<Self> {
+        let bitmap_bytes = batch.column(0).as_binary::<i32>().value(0);
+        let bitmap =
+            RoaringTreemap::deserialize_from(bitmap_bytes).map_err(|e| Error::Internal {
+                message: format!("Error deserializing ngram list: {}", e),
+                location: location!(),
+            })?;
+        Ok(Self { bitmap })
+    }
+
+    fn intersect<'a>(lists: impl IntoIterator<Item = &'a Self>) -> RoaringTreemap {
+        let mut iter = lists.into_iter();
+        let mut result = iter
+            .next()
+            .map(|list| list.bitmap.clone())
+            .unwrap_or_default();
+        for list in iter {
+            result &= &list.bitmap;
+        }
+        result
+    }
+}
+
+/// Reads on-demand ngram posting lists from storage (and stores them in a cache)
+struct NGramPostingListReader {
+    reader: Arc<dyn IndexReader>,
+    cache: Cache<u32, Arc<NGramPostingList>>,
+}
+
+impl DeepSizeOf for NGramPostingListReader {
+    fn deep_size_of_children(&self, _: &mut deepsize::Context) -> usize {
+        self.cache.weighted_size() as usize
```

Comment:
Why is this the correct answer? Wouldn't it be 4 bytes + Size of `NGramPostingList` per entry?

TBH it's unclear what weighted size does.

Reply:
When we define the cache we say:

```
cache: Cache::builder()
    .max_capacity(*CACHE_SIZE as u64)
    .weigher(|_, posting: &Arc<NGramPostingList>| posting.deep_size_of() as u32)
    .build(),
```

So the `weigher` is set to `deep_size_of` for the `NGramPostingList`.  In other words, this is a bytes-based cache (not an entry-based cache).  As a result, the `weighted_size` _should_ be a fast approximation of the total index size (I may be wrong here but this is my interpretation)

Reply:
Oh I missed the weigher. That is cool!

Reply:
Yes, but I can't take credit for it :laughing: .  I stole this pattern from @BubbleCal :laughing: